[testing] Shared Behaviors best practices

The testing packages use Mocha, which [discourages](https://mochajs.org/#arrow-functions) passing arrow functions.

**When creating [shared behaviors](https://github.com/mochajs/mocha/wiki/Shared-Behaviours
), using the Mocha Context seems mandatory**. Devs need to rewrite all their callbacks using function statements instead of arrows as soon as they want to share some test behaviors between files.

Therefore, **unless we find some other way to write shared behaviors** with ES modules, and share the related recommendations, it may be a good idea to use the function statements everywhere in the example tests.

https://github.com/open-wc/open-wc/blob/master/packages/testing/README.md#example-tests

Here is an example:
:point_right: https://github.com/noelmace/devcards/blob/test/test/open-wc/components/card-stack.test.js

Related external issues & documentations:
https://github.com/meteor/guide/issues/318
https://github.com/mochajs/mocha/issues/1856

If you can confirm that this is an issue, I'm willing to work on this myself

Thx @daKmoR but like I said previously, even if I would prefer this solution too, it doesn't work, because `shouldBehaveLikeAUser` is called after `beforeEach`, but before the callback of beforeEach. So `user` is undefined in `shouldBehaveLikeAUser`.

In order to be sure, I reproduced the exact same code in my project: https://github.com/noelmace/devcards/commit/0908d058e3b6c121236f26b5a21c63b32ef748ee

Here is the output:

```
npm run vanilla-test -- --verbose=false

> flashcards-app@0.0.0 vanilla-test /home/noel/git/web-on-fire/flashcards-app
> karma start karma.owc.conf.js --coverage --grep 'test/vanilla-chai/**/*.test.js' "--verbose=false"


START:
16 08 2019 13:27:24.357:WARN [filelist]: Pattern "/home/noel/git/web-on-fire/flashcards-app/__snapshots__/**/*.md" does not match any file.
16 08 2019 13:27:24.369:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
16 08 2019 13:27:24.370:INFO [launcher]: Launching browsers ChromeHeadlessNoSandbox with concurrency unlimited
16 08 2019 13:27:24.373:INFO [launcher]: Starting browser ChromeHeadless
16 08 2019 13:27:24.655:INFO [HeadlessChrome 76.0.3809 (Linux 0.0.0)]: Connected on socket fgpyQ4sLKKTU_K_GAAAA with id 41202211
  by daKmoR
    User
      ✖ should have .name.first
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR User should have .name.first FAILED
        TypeError: Cannot read property 'name' of undefined
            at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:3:10)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
      ✖ should have .name.last
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR User should have .name.last FAILED
        TypeError: Cannot read property 'name' of undefined
            at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:7:10)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
      .fullname()
        ✖ should return the full name
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR User .fullname() should return the full name FAILED
        TypeError: Cannot read property 'fullname' of undefined
            at Context.fullname (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:12:12)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
    Admin
      ✖ should have .name.first
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR Admin should have .name.first FAILED
        TypeError: Cannot read property 'name' of undefined
            at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:3:10)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
      ✖ should have .name.last
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR Admin should have .name.last FAILED
        TypeError: Cannot read property 'name' of undefined
            at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:7:10)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
      ✖ should be an .admin
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR Admin should be an .admin FAILED
        TypeError: Cannot read property 'be' of undefined
            at Context.be (test/vanilla-chai/shared-behaviors/dakmor/user.test.js:23:25)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)
      .fullname()
        ✖ should return the full name
HeadlessChrome 76.0.3809 (Linux 0.0.0) by daKmoR Admin .fullname() should return the full name FAILED
        TypeError: Cannot read property 'fullname' of undefined
            at Context.fullname (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:12:12)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

Finished in 0.045 secs / 0 secs @ 13:27:26 GMT+0200 (GMT+02:00)

SUMMARY:
✔ 0 tests completed
✖ 7 tests failed

FAILED TESTS:
  by daKmoR
    User
      ✖ should have .name.first
        HeadlessChrome 76.0.3809 (Linux 0.0.0)
      TypeError: Cannot read property 'name' of undefined
          at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:3:10)
          at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

      ✖ should have .name.last
        HeadlessChrome 76.0.3809 (Linux 0.0.0)
      TypeError: Cannot read property 'name' of undefined
          at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:7:10)
          at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

      .fullname()
        ✖ should return the full name
          HeadlessChrome 76.0.3809 (Linux 0.0.0)
        TypeError: Cannot read property 'fullname' of undefined
            at Context.fullname (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:12:12)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

    Admin
      ✖ should have .name.first
        HeadlessChrome 76.0.3809 (Linux 0.0.0)
      TypeError: Cannot read property 'name' of undefined
          at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:3:10)
          at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

      ✖ should have .name.last
        HeadlessChrome 76.0.3809 (Linux 0.0.0)
      TypeError: Cannot read property 'name' of undefined
          at Context.name (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:7:10)
          at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

      ✖ should be an .admin
        HeadlessChrome 76.0.3809 (Linux 0.0.0)
      TypeError: Cannot read property 'be' of undefined
          at Context.be (test/vanilla-chai/shared-behaviors/dakmor/user.test.js:23:25)
          at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)

      .fullname()
        ✖ should return the full name
          HeadlessChrome 76.0.3809 (Linux 0.0.0)
        TypeError: Cannot read property 'fullname' of undefined
            at Context.fullname (test/vanilla-chai/shared-behaviors/dakmor/helpers.js:12:12)
            at Test.apply [as run] (node_modules/karma-mocha-snapshot/lib/adapter.js:14:17)


=============================== Coverage summary ===============================
Statements   : 4.55% ( 2/44 )
Branches     : 0% ( 0/7 )
Functions    : 0% ( 0/6 )
Lines        : 4.76% ( 2/42 )
================================================================================
16 08 2019 13:27:26.510:ERROR [reporter.coverage-istanbul]: Coverage for statements (4.55%) does not meet global threshold (80%)
16 08 2019 13:27:26.510:ERROR [reporter.coverage-istanbul]: Coverage for lines (4.76%) does not meet global threshold (80%)
16 08 2019 13:27:26.510:ERROR [reporter.coverage-istanbul]: Coverage for branches (0%) does not meet global threshold (80%)
16 08 2019 13:27:26.510:ERROR [reporter.coverage-istanbul]: Coverage for functions (0%) does not meet global threshold (80%)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! flashcards-app@0.0.0 vanilla-test: `karma start karma.owc.conf.js --coverage --grep 'test/vanilla-chai/**/*.test.js' "--verbose=false"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the flashcards-app@0.0.0 vanilla-test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/noel/.npm/_logs/2019-08-16T11_27_26_541Z-debug.log
```

--- a/docs/about/blog.md
+++ b/docs/about/blog.md
@@ -19,3 +19,7 @@ by [Westbrook Johnson](https://github.com/westbrook)
 ## [Web Components: spoils of the browser wars](https://www.andrewzigler.com/blog/2019/02/28/web-components-spoils-of-the-browser-wars/)
 
 by [Andrew Zigler](https://github.com/azigler)
+
+## [Shared Behaviors best practices with Mocha](https://dev.to/open-wc/shared-behaviors-best-practices-with-mocha-519d)
+
+by [Noël Macé](https://github.com/noelmace)

